glusterd2/commands/volumes/volume-profile.go

```diff
@@ -0,0 +1,309 @@
+		stats.StatsInfo[key]["%-latency"] = fmt.Sprintf("%f", tmpPercentageAvgLatencyForEachFop)
+	}
+	stats.PercentageAvgLatency = tmpPercentageAvgLatency
+
+	return stats
+}
+
+// populateStatsWithFop populates both interval/cumulative stats with fops and its profile info
+func populateStatsWithFop(stats StatType, key string, value string, ktype keyType) StatType {
+	var fop string
+	var k string
+	// Decode the fop and stat for the given key
+	switch ktype {
+	case cumulativeType:
+		fop, k = decodeCumulativeKey(key)
+		break
+	case intervalType:
+		fop, k = decodeIntervalKey(key)
+		break
+	}
+	if fop != "" && fop != "NULL" {
+		if _, ok := stats.StatsInfo[fop]; ok {
+			stats.StatsInfo[fop][k] = value
+		} else {
+			// Create new map with fop as key if this particular Fop is encountered for the first time
+			stats.StatsInfo[fop] = make(map[string]string)
+			stats.StatsInfo[fop][k] = value
+		}
+	} else {
+		if k == "read" {
+			stats.DataRead = value
+		} else if k == "write" {
+			stats.DataWrite = value
+		} else if k == "duration" {
+			stats.Duration = value
+		}
+	}
+	return stats
+}
+
+// Decode key for fop and stat. keys starting with -1 belong to cumulative stat
+// Eg: -1-12-maxlatency fop : 12        stat: maxlatency
+// -1-duration          fop : ""        stat: duration
+func decodeCumulativeKey(key string) (string, string) {
+	var fop string
+	var k string
+	s := strings.Split(key, "-")
+	if len(s) == 4 {
+		k = s[len(s)-1]
+		index, _ := strconv.Atoi(s[2])
+		fop = fops[index]
+	} else if len(s) == 3 {
+		k = s[len(s)-1]
+		fop = ""
+	}
+	return fop, k
+}
+
+// Decode key for fop and stat for keys belonging to interval stat
+// Eg: 12-12-maxlatency  IntervalNo.: 12  fop : 12        stat: maxlatency
+// 12-duration           IntervalNo.: 12  fop : ""        stat: duration
+func decodeIntervalKey(key string) (string, string) {
+	var fop string
+	var k string
+	s := strings.Split(key, "-")
+	if len(s) == 3 {
+		k = s[len(s)-1]
+		index, _ := strconv.Atoi(s[1])
+		fop = fops[index]
+	} else if len(s) == 2 {
+		k = s[len(s)-1]
+		fop = ""
+	}
+	return fop, k
+}
+
+func txnVolumeProfile(c transaction.TxnCtx) error {
+	var volinfo volume.Volinfo
+	var nodeProfileInfo []map[string]string
+	if err := c.Get("volinfo", &volinfo); err != nil {
+		return err
+	}
+
+	var option string
+	if err := c.Get("option", &option); err != nil {
+		return err
+	}
+
+	for _, b := range volinfo.GetLocalBricks() {
+		brickDaemon, err := brick.NewGlusterfsd(b)
+		if err != nil {
+			return err
+		}
+
+		c.Logger().WithField(
+			"volume", volinfo.Name).Info("Starting volume profile operation")
+
+		client, err := daemon.GetRPCClient(brickDaemon)
+		if err != nil {
+			c.Logger().WithError(err).WithField(
+				"brick", b.String()).Error("failed to connect to brick, sending SIGTERM")
+			return err
+		}
+		reqDict := make(map[string]string)
+		switch option {
+		case "info":
+			reqDict["peek"] = "0"
+			reqDict["op"] = "3"
```

Comment:
too much of hard coded numbers. If I assume these numbers are coming from a list, can they be made enums?

Reply:
@atinmu These numbers are what needs to be sent in the RPC request to denote which type of profile info user wants i.e peek, cumulative, incremental etc. I guess I can use enums for this.